src/cms/logic.py

```diff
@@ -1,15 +1,367 @@
 __copyright__ = "Copyright 2023 Birkbeck, University of London"
-__author__ = "Joseph Muller"
+__author__ = "Open Library of Humanities"
 __license__ = "AGPL v3"
 __maintainer__ = "Open Library of Humanities"
 
 import os
+import json
 import glob
+import json
+import time
+import requests
+from typing import Tuple, List, Dict
+from urllib.parse import urljoin, urlparse
+from bs4 import BeautifulSoup, Comment, NavigableString, Tag
 
 from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured
+from django.core.files.base import ContentFile
 
 from cms import models as models
+from journal import models as journal_models
+from press import models as press_models
+from repository import models as repository_models
 from utils import setting_handler
+from utils.function_cache import cache
+from utils.logger import get_logger
+
+logger = get_logger(__name__)
+
+SITE_SEARCH_PATH = os.path.join(
+    settings.MEDIA_ROOT,
+    'press',
+    settings.SITE_SEARCH_DIR,
+)
+
+
+def add_search_index_document(
+    docs: List[Dict],
+    url: str,
+    name: str,
+    text: str,
+) -> List[Dict]:
+    """
+    Adds the data for a part or page to the search index
+    """
+    data = {}
+    data['id'] = len(docs)
+    data['url'] = url
+    data['name'] = name
+    data['text'] = text
+    docs.append(data)
+    return docs
+
+
+def gobble_sibling_text(
+    sibling: Tag | NavigableString,
+    original_part: Tag | NavigableString,
+) -> str:
+    """
+    Recursively collects a series of parallel elements for
+    indexing together, stopping before another heading
+    """
+
+    # Handle no sibling
+    if not sibling:
+        return ''
+
+    # Handle another heading
+    if sibling.name == original_part.name:
+        return ''
+    if isinstance(sibling, Tag) and sibling.find(original_part.name):
+        return ''
+
+    # Recursively get sibling text
+    if isinstance(sibling, NavigableString):
+        sibling_text = sibling.strip()
+    else:
+        sibling_text = sibling.get_text().strip()
+    next_sibling_text = gobble_sibling_text(
+        sibling.next_sibling,
+        original_part
+    )
+
+    # Decompose
+    if isinstance(sibling, NavigableString):
+        sibling.replace_with('')
+    else:
+        sibling.decompose()
+
+    gobbled_text = sibling_text + ' ' + next_sibling_text
+    return gobbled_text.strip()
+
+
+def get_text_for_parent(
+    parent: Tag | NavigableString,
+    original_part: Tag | NavigableString,
+) -> str:
+    if not parent:
+        return ''
+
+    text = gobble_sibling_text(parent.next_sibling, original_part)
+    if len(text.strip()) > len(original_part.get_text().strip()):
+        return text
+    else:
+        return get_text_for_parent(parent.parent, original_part)
+
+
+def get_text_for_header(part: Tag | NavigableString) -> str:
```

Comment:
While I'm a big fan of type hints, I don't think we are ready to include them in the Janeway repository just yet, since most of the codebase is still using Google style docstring annotations (*"consistency is key"*)
Also, this operand is only compatible with python 3.10+ and we still support 3.8+

Reply:
OK, I have used docstrings until we can upgrade to type annotations